python/pyarrow/_dataset.pyx

```diff
@@ -2505,6 +2505,43 @@ cdef class Partitioning(_Weakrefable):
         result = self.partitioning.Parse(tobytes(path))
         return Expression.wrap(GetResultValue(result))
 
+    def format(self, expr):
+        """
+        Convert a filter expression into a tuple of (directory, filename) using 
+        the current partitioning scheme
+
+        Parameters
+        ----------
+        expr : pyarrow.dataset.Expression
+
+        Returns
+        -------
+        tuple[str, str]
+
+        Examples
+        --------
+
+        Specify the Schema for paths like "/2009/June":
+
+        >>> import pyarrow as pa
+        >>> import pyarrow.dataset as ds
+        >>> import pyarrow.compute as pc
+        >>> part = ds.partitioning(pa.schema([("year", pa.int16()),
+        ...                                   ("month", pa.string())]))
+        >>> part.format(
+        ...     (pc.field("year") == 1862) & (pc.field("month") == "Jan")
+        ... )
+        """
+        cdef:
+            CResult[CPartitionPathFormat] result
```

Comment:
We can optimize the temporal variable `result` by combine `Format` and `GetResultValue`

Reply:
> Wonder if this is true in CPython compiler? Since this style is widely used in `pyx` scripts...
It's ok. This style is also widely used in pyx file like array.pxi  and it can save one copy of `T`

Reply:
@pitrou I'm not familiar with CPython, and I didn't find the answer, do you think this would be cost, or this separate assign could be optimized out by compiler?

Reply:
This would simply be more readable by removing the temporary.

Reply:
Removed in [`4b4f3cf` (#43740)](https://github.com/apache/arrow/pull/43740/commits/4b4f3cfa39ec11c05d6fe779d7555f953f4ff88d).